File: 024-二叉搜索树的后序遍历序列/024-VerifySquenceOfBST.py
# -*- coding:utf-8 -*-
class Solution:
    def VerifySquenceOfBST(self, sequence):
        # write code here
        if len(sequence)==0:
            return False
        return self.verify(sequence)
    def verify(self,sequence):
        if len(sequence)<=1:
            return True
        idx=-1
        for i,val in enumerate(sequence[:-1]):
            if val >sequence[-1]:
                idx=i
                break
        if idx==-1:
            return True
        for i in range(idx,len(sequence)-1):
            if sequence[i]<sequence[-1]:
                return False
        return self.verify(sequence[:idx]) and self.verify(sequence[idx:-1])

File: 024-二叉搜索树的后序遍历序列/024-VerifySquenceOfBST.cpp
class Solution {
public:
    bool VerifySquenceOfBST(vector<int> sequence) {
        if(sequence.size()==0){
            return false;
        }
        return verify(sequence,0,sequence.size()-1);
    }
    bool verify(vector<int> &sequence,int start,int end) {
        if(start>=end){
            return true;
        }
        int idx=-1;
        for(int i=start;i<end;i++){
            if(sequence[i]>sequence[end])
            {
                idx=i;
                break;
            }
        }
        if(idx==-1){
            return true;
        }
        for(int i=idx;i<end;i++){
            if(sequence[i]<sequence[end]){
                return false;
            }
        }
        return verify(sequence,start,idx-1) && verify(sequence,idx,end-1);
        
    }
};
